Refactor: AsyncIO no IoT Service

djoseph-png opened the pull request:
Resumo

Convertidos os métodos de HueLightDevice, SmartSpeakerDevice e SmartToiletDevice para assíncronos (async def + await asyncio.sleep).

Atualizado IOTService para trabalhar com métodos assíncronos (connect, disconnect, send_message).

Substituída a execução sequencial dos programas por combinação de paralelismo e sequência usando run_parallel e run_sequence.

Mantida a ordem lógica de dependências (ex.: SWITCH_ON antes de PLAY_SONG, FLUSH antes de CLEAN).

Ganho de desempenho: execução caiu de ~5.0s para ~2.5s.

Evidências
Antes (main.py sequencial)

Depois (main.py assíncrono)

Notas

Testado com flake8 → sem erros após ajustes (E501, E704).

Mantida compatibilidade com a API original de mensagens (Message, MessageType).

Estrutura pronta para expandir facilmente novos devices e programas assíncronos.
<img width="1905" height="1119" alt="run-before-asyncio" src="https://github.com/user-attachments/assets/e0bd3bde-5c0f-410a-88b2-b8f641116b33" />
<img width="1853" height="833" alt="run-after-asyncio" src="https://github.com/user-attachments/assets/676dd7c2-c709-432b-828b-252e4622e9ee" />

djoseph-png committed:
Add baseline run screenshot

djoseph-png committed:
Async refactor: devices+service async, parallel connect, run_parallel/run_sequence; add screenshots

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,3 +1,4 @@
+# app/iot/service.py
```
This top-of-file marker is a good place to note: the file currently lacks the required orchestration helpers and program composition. The task description requires adding run_sequence and run_parallel and refactoring wake_up/sleep orchestration (first wake_up, then sleep). Add those helpers and the orchestration logic here or in a coordinator module and remove any prior program-variable-based scheduling .

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -9,41 +10,35 @@ def generate_id(length: int = 8) -> str:
 class Device(Protocol):
-    def connect(self) -> None:
-        ...  # Ellipsis - similar to "pass", but sometimes has different meaning
+    async def connect(self) -> None:
```
The Device protocol declares async connect/disconnect/send_message — this matches the task requirement that these interfaces be async (requirement 1.1). Good: device implementations already use asyncio.sleep for non-blocking delays, so the device layer is compliant .

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -9,41 +10,35 @@ def generate_id(length: int = 8) -> str:
 
-    def register_device(self, device: Device) -> str:
-        device.connect()
+    async def register_device(self, device: Device) -> str:
+        await device.connect()
```
register_device awaits device.connect and then registers the device. That method itself is correct. However, to satisfy the task's concurrency requirement you must call register_device for multiple devices concurrently using asyncio.gather from the orchestration layer — otherwise devices will still register sequentially. Implement an orchestration that gathers multiple register_device coroutines (requirement 3.2).

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -9,41 +10,35 @@ def generate_id(length: int = 8) -> str:
-    def send_msg(self, msg: Message) -> None:
-        self.devices[msg.device_id].send_message(msg.msg_type, msg.data)
+    async def send_msg(self, msg: Message) -> None:
+        await self.devices[msg.device_id].send_message(msg.msg_type, msg.data)
```
send_msg correctly awaits the device's send_message. Make sure your orchestration ensures devices are connected before calling send_msg (i.e., await all register_device gather results first) and ensure you await unregister_device/disconnect at the end of the whole flow. Also implement run_sequence/run_parallel so message sending for wake_up/sleep respects dependencies and parallelism (requirements 1.4, 3.7).

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision:** REQUEST_CHANGES — core orchestration requirements are missing, which blocks task completion.

**Critical blockers to fix:**
- Missing helpers and orchestration (requirement 2.2, 3.3): In `app/iot/service.py` top section, there is no `run_sequence` or `run_parallel`, and the wake-up/sleep flows are not refactored to use them. This is explicitly required and is the core of the task. Add:
  - `async def run_sequence(*coros):` iterate and `await` each in order.
  - `async def run_parallel(*coros):` `await asyncio.gather(*coros)`.
  - Replace any previous program-variable-based scheduling with explicit `run_sequence`/`run_parallel` composition, first implementing the wake_up flow, then the sleep flow, preserving dependencies (e.g., switch-on before play; flush before clean).
- Missing concurrent registration (requirement 3.2): While `register_device` is correct, you must orchestrate registering multiple devices concurrently using `asyncio.gather`. Ensure you `await` all registrations before sending any messages. Place this orchestration in `service.py` or a coordinator module.
- Missing end-to-end flow (requirement 3.7): Ensure the orchestration follows `connect -> actions -> disconnect`. After programs complete, `await unregister_device`/`disconnect` for all devices.

**Specific code locations and guidance:**
- `app/iot/service.py` (top of file): Add `run_sequence` and `run_parallel` helpers and the orchestration that composes wake_up then sleep. Remove any prior program-variable-based scheduling if present.
- `app/iot/service.py` (or coordinator): Use `asyncio.gather(*(service.register_device(d) for d in devices))` to register concurrently; later, gather `unregister_device` calls.

**What’s good:**
- Device protocol and implementations are async and non-blocking; `register_device`, `unregister_device`, and `send_msg` correctly `await` device operations. This foundation is solid.

**Minor improvements (non-blockers):**
- Make `unregister_device` safer by popping the device before disconnecting to avoid `KeyError`.
- Align `Device.send_message` signature default with implementations.

You’re very close — add the orchestration helpers and compose the wake_up/sleep flows with proper parallelism and ordering, and this will meet the requirements.

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

djoseph-png committed:
Async orchestration: helpers in service.py, parallel register/unregister, wake/sleep with run_parallel+run_sequence; flake8 ok

djoseph-png commented:
Changes applied per review:

Added run_sequence and run_parallel helpers in app/iot/service.py.

Centralized orchestration in service.py via orchestrate():

concurrent register_device with asyncio.gather (req. 3.2),

wake-up and sleep flows using run_parallel + run_sequence preserving dependencies (req. 2.2, 3.3),

end-to-end connect → actions → disconnect with concurrent unregister_device (req. 3.7).

Improved unregister_device safety using pop.

Aligned Device.send_message signature default (data: str = "").

All files pass flake8 (E501/E704/VNE001 addressed).
Evidence: screenshots/run-before-asyncio.png and screenshots/run-after-asyncio2
<img width="1782" height="974" alt="run-after-asyncio2" src="https://github.com/user-attachments/assets/cbdc9bf4-6f4a-48dc-8175-c0d6d75fb074" />
.png.

mateacademy-ai-mentor commented on app/iot/devices.py:
```diff
@@ -1,64 +1,72 @@
-import time
+# app/iot/devices.py
+import asyncio
```
Good: `asyncio` is imported. This is required so the device methods can use `asyncio.sleep` and be non-blocking per the task requirements.

mateacademy-ai-mentor commented on app/iot/devices.py:
```diff
@@ -1,64 +1,72 @@
 class HueLightDevice:
-    def connect(self) -> None:
+    async def connect(self) -> None:
         print("Connecting Hue Light.")
-        time.sleep(TIME_TO_SLEEP)
+        await asyncio.sleep(TIME_TO_SLEEP)
```
HueLightDevice.connect is declared `async` and awaits `asyncio.sleep`, which satisfies the requirement that `connect` be asynchronous and non-blocking. This allows registration to run concurrently in the orchestrator.

mateacademy-ai-mentor commented on app/iot/devices.py:
```diff
@@ -1,64 +1,72 @@
-    def send_message(self, message_type: MessageType, data: str = "") -> None:
+    async def send_message(
+        self, message_type: MessageType, data: str = ""
+    ) -> None:
         print(
-            f"Hue Light handling message of type {message_type.name} with data [{data}]."
+            f"Hue Light handling message of type {message_type.name} "
+            f"with data [{data}]."
         )
-        time.sleep(TIME_TO_SLEEP)
+        await asyncio.sleep(TIME_TO_SLEEP)
```
HueLightDevice.send_message has the expected signature (`message_type: MessageType, data: str = ""`) and is `async`. Using `asyncio.sleep` here ensures sending messages won't block other concurrent operations.

mateacademy-ai-mentor commented on app/iot/devices.py:
```diff
@@ -1,64 +1,72 @@
+        await asyncio.sleep(TIME_TO_SLEEP)
         print("Smart Speaker connected.")
 
-    def disconnect(self) -> None:
+    async def disconnect(self) -> None:
         print("Disconnecting Smart Speaker.")
-        time.sleep(TIME_TO_SLEEP)
+        await asyncio.sleep(TIME_TO_SLEEP)
         print("Smart Speaker disconnected.")
 
-    def send_message(self, message_type: MessageType, data: str = "") -> None:
+    async def send_message(
+        self, message_type: MessageType, data: str = ""
+    ) -> None:
         print(
-            f"Smart Speaker handling message of type {message_type.name} with data [{data}]."
+            f"Smart Speaker handling message of type {message_type.name} "
+            f"with data [{data}]."
         )
-        time.sleep(TIME_TO_SLEEP)
+        await asyncio.sleep(TIME_TO_SLEEP)
```
SmartSpeakerDevice methods are correctly async and use `asyncio.sleep`. Note that the speaker's `send_message` matches the expected signature and will cooperate with the orchestrator's sequencing/gathering.

mateacademy-ai-mentor commented on app/iot/devices.py:
```diff
@@ -1,64 +1,72 @@
+        await asyncio.sleep(TIME_TO_SLEEP)
         print("Smart Toilet connected.")
 
-    def disconnect(self) -> None:
+    async def disconnect(self) -> None:
         print("Disconnecting Smart Toilet.")
-        time.sleep(TIME_TO_SLEEP)
+        await asyncio.sleep(TIME_TO_SLEEP)
         print("Smart Toilet disconnected.")
 
-    def send_message(self, message_type: MessageType, data: str = "") -> None:
+    async def send_message(
+        self, message_type: MessageType, data: str = ""
+    ) -> None:
         print(
-            f"Smart Toilet handling message of type {message_type.name} with data [{data}]."
+            f"Smart Toilet handling message of type {message_type.name} "
+            f"with data [{data}]."
         )
-        time.sleep(TIME_TO_SLEEP)
+        await asyncio.sleep(TIME_TO_SLEEP)
```
SmartToiletDevice methods are also async and use `asyncio.sleep`. The toilet's `send_message` will allow the orchestrator to implement `FLUSH` -> `CLEAN` sequencing without blocking other tasks.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,113 @@
+import asyncio
 import random
 import string
 from typing import Protocol
```
Import typing additions are required by the task. Currently you have `from typing import Protocol`. The assignment asks the orchestration helpers to accept `Awaitable[Any]`. Please import `Any` and `Awaitable` and use them in helper signatures so static typing matches the guidance: `from typing import Any, Awaitable` (or add them to the existing import). This change is requested by the description and checklist.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,113 @@
+# Helpers de orquestração
+async def run_sequence(*coros) -> None:
+    for coro in coros:
+        await coro
```
The `run_sequence` helper is implemented correctly functionally, but it lacks the typing requested in the task. Change its signature to accept `Awaitable[Any]` (for example: `async def run_sequence(*functions: Awaitable[Any]) -> None:`) so it matches the provided guidance and typing constraints. This addresses requirement 2.7 / 3.7 in the checklist.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,113 @@
+
+
+async def run_parallel(*coros) -> None:
+    await asyncio.gather(*coros)
```
The `run_parallel` helper should also include typing on its parameters (e.g., `async def run_parallel(*functions: Awaitable[Any]) -> None:`). This keeps the helpers consistent with the assignment snippet and avoids type confusion.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,113 @@
+                )
+            ),
+        ),
+    )
```
The wake-up orchestration currently uses an outer `run_parallel` that contains a `run_sequence` for the speaker. Functionally this preserves speaker ordering, but the assignment specifically requests composing the program as stages using `run_sequence` with inner `run_parallel` stage calls (so multi-stage ordering across devices is explicit). Please refactor the wake-up block into an outer `run_sequence` of stage-level `run_parallel` calls. Example stage breakdown:

- Stage 1 (run_parallel): SWITCH_ON for Hue and Speaker
- Stage 2 (run_parallel): PLAY_SONG for Speaker

This ensures stage ordering is explicit and matches the task description.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,113 @@
+            service.send_msg(Message(toi_id, MessageType.FLUSH)),
+            service.send_msg(Message(toi_id, MessageType.CLEAN)),
+        ),
+    )
```
The sleep orchestration should be refactored similarly into an outer `run_sequence` containing stage-level `run_parallel` calls. Ensure dependencies remain: the toilet must `FLUSH` then `CLEAN` (keep that pair sequenced using `run_sequence` inside the relevant stage). One possible structure:

- Stage 1 (run_parallel): SWITCH_OFF for Hue and Speaker
- Stage 2 (run_parallel): `run_sequence(Message(TOILET, FLUSH), Message(TOILET, CLEAN))` (so toilet's flush+clean runs after switch-offs stage, or in whichever stage ordering you intend)

Refactor the current top-level `run_parallel` to this stage-based composition to satisfy the orchestration shape requirement.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,113 @@
-            self.send_msg(msg)
-        print("=====END OF PROGRAM======")
+    async def send_msg(self, msg: Message) -> None:
+        await self.devices[msg.device_id].send_message(msg.msg_type, msg.data)
```
Consider renaming `IOTService.send_msg` to `send_message` to align naming with device `send_message` and the task description. This is optional but will reduce confusion between device-level method names and the service API. If you keep `send_msg`, make sure it's acceptable as the public API name for the project.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,113 @@
+    toi = SmartToiletDevice()
+
+    # registrar devices em paralelo
+    hue_id, spk_id, toi_id = await asyncio.gather(
+        service.register_device(hue),
+        service.register_device(spk),
+        service.register_device(toi),
+    )
```
Good: device registration is done concurrently using `asyncio.gather(...)` and the returned IDs are awaited before running the programs. This satisfies the requirement to register devices in parallel. Keep this pattern.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,113 @@
-    def unregister_device(self, device_id: str) -> None:
-        self.devices[device_id].disconnect()
-        del self.devices[device_id]
+    async def unregister_device(self, device_id: str) -> None:
+        device = self.devices.pop(device_id, None)
+        if device is not None:
+            await device.disconnect()
```
Good: `unregister_device` uses `pop` then checks for `None` before disconnecting. This prevents a potential `KeyError` and is the safer pattern for unregistering devices. Keep this behavior.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision: REQUEST_CHANGES**

Your async device implementations look great: all device methods are `async`, use `asyncio.sleep`, and the orchestrator registers/unregisters devices concurrently. However, there are two required items from the assignment/checklist that need updates before approval:

- Critical blockers to fix:
  1) Helper typing (app/iot/service.py): `run_sequence` and `run_parallel` must accept `Awaitable[Any]` as per the task. Add `from typing import Any, Awaitable` and update signatures to `async def run_sequence(*functions: Awaitable[Any]) -> None:` and `async def run_parallel(*functions: Awaitable[Any]) -> None:`. This is an explicit requirement (2.7/3.7) and currently missing.
  2) Orchestration shape (app/iot/service.py): Refactor wake-up and sleep into an outer `run_sequence` of stage-level `run_parallel` calls (explicit staging). Example for wake-up: Stage 1 `run_parallel` -> `SWITCH_ON` Hue + Speaker; Stage 2 `run_parallel` -> `PLAY_SONG` Speaker. For sleep, ensure toilet `FLUSH -> CLEAN` remains a `run_sequence` inside the appropriate stage. This aligns with the assignment’s required composition pattern.

- Minor suggestions (non-blockers):
  - Consider renaming `IOTService.send_msg` to `send_message` for API consistency, or document the naming.  
  - Make `send_msg` handle unknown device IDs gracefully (e.g., use `get` and raise a clear `ValueError`).

Once you add the helper typing and refactor the orchestration into explicit stages, the submission will fully meet the requirements. Nice work on the async conversion and concurrency so far!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

djoseph-png commented:
Changes applied per review

Helper typing (req. 2.7 / 3.7)
Added from typing import Any, Awaitable and updated helpers:

async def run_sequence(*functions: Awaitable[Any]) -> None: ...
async def run_parallel(*functions: Awaitable[Any]) -> None: ...


Explicit orchestration shape (req. 3.3)
Refactored orchestration into an outer run_sequence of stage-level run_parallel calls:

Wake-up

Stage 1 → run_parallel: SWITCH_ON Hue + Speaker

Stage 2 → run_parallel: PLAY_SONG Speaker

Sleep

Stage 1 → run_parallel: SWITCH_OFF Hue + Speaker

Stage 2 → run_sequence: FLUSH → CLEAN Toilet

Concurrent registration/unregistration (req. 3.2 / 3.7)
asyncio.gather used to register and unregister all devices in parallel.

Minor improvements

IOTService.send_msg now checks device_id and raises a clear ValueError if unknown.

Device protocol signature aligned (data: str = "").

✅ All files pass flake8.
✅ Async devices + orchestrator verified with before/after screenshots.
<img width="1845" height="992" alt="run-after-asyncio3" src="https://github.com/user-attachments/assets/c8ec529f-6997-494c-905f-4c373d765ecd" />

djoseph-png committed:
Add async orchestration: typed helpers, staged wake/sleep flows, parallel register/unregister (flake8 ok)

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,122 @@
 
-# Protocol is very similar to ABC, but uses duck typing
-# so devices should not inherit for it (if it walks like a duck, and quacks like a duck, it's a duck)
+# Helpers com typing (Awaitable[Any]) — requisito 2.7 / 3.7
+async def run_sequence(*functions: Awaitable[Any]) -> None:
+    for func in functions:
+        await func
```
Good: run_sequence is typed to accept Awaitable[Any] and properly awaits each awaitable in order, which preserves sequential dependencies as required. Consider adding a short docstring or renaming the parameter from `functions` to `awaitables` to make the expected input clearer to readers.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,122 @@
+        await func
+
+
+async def run_parallel(*functions: Awaitable[Any]) -> None:
+    await asyncio.gather(*functions)
```
Good: run_parallel uses asyncio.gather to run awaitables concurrently as required. Note (optional): if you want to surface all exceptions rather than cancelling on the first, you can consider `return_exceptions=True` depending on desired error semantics — but the current default behavior is acceptable for this task.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,122 @@
-        for msg in program:
-            self.send_msg(msg)
-        print("=====END OF PROGRAM======")
+    async def send_msg(self, msg: Message) -> None:
+        device = self.devices.get(msg.device_id)
+        if not device:
+            raise ValueError(f"Device {msg.device_id} not found")
+        await device.send_message(msg.msg_type, msg.data)
```
Suggestion: `send_msg` currently implements the service->device send. For API consistency with the Device protocol and device implementations (which use `send_message`), consider renaming this method to `send_message` or add a comment documenting the naming to avoid confusion. Also, the method raises ValueError when a device is not found — that's fine, but consider whether a more specific exception or a clearer message is desired.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,122 @@
+    toi = SmartToiletDevice()
+
+    # registro concorrente
+    hue_id, spk_id, toi_id = await asyncio.gather(
+        service.register_device(hue),
+        service.register_device(spk),
+        service.register_device(toi),
+    )
```
Good: devices are registered concurrently via `asyncio.gather(...)`, which satisfies the concurrent registration requirement. Note that `register_device` still awaits `device.connect()` internally (correct) so the concurrency is achieved by calling register_device concurrently, not by making connect non-blocking — which is the expected behavior.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,122 @@
+
+    # ===== WAKE-UP program =====
+    print("=====RUNNING PROGRAM======")
+    await run_sequence(
+        # Stage 1: ligar devices em paralelo
+        run_parallel(
+            service.send_msg(Message(hue_id, MessageType.SWITCH_ON)),
+            service.send_msg(Message(spk_id, MessageType.SWITCH_ON)),
+        ),
+        # Stage 2: tocar música
+        run_parallel(
+            service.send_msg(
+                Message(
+                    spk_id,
+                    MessageType.PLAY_SONG,
+                    "Rick Astley - Never Gonna Give You Up",
+                )
+            ),
+        ),
+    )
+    print("=====END OF PROGRAM======")
```
Good: Wake-up program is composed as an outer `run_sequence` of stage-level calls; stage 1 turns on devices in parallel and stage 2 plays the song. This ensures the Speaker is switched on before `PLAY_SONG`, preserving logical order while maximizing parallelism where possible.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,122 @@
+
+    # ===== SLEEP program =====
+    print("=====RUNNING PROGRAM======")
+    await run_sequence(
+        # Stage 1: desligar luz + speaker (em paralelo)
+        run_parallel(
+            service.send_msg(Message(hue_id, MessageType.SWITCH_OFF)),
+            service.send_msg(Message(spk_id, MessageType.SWITCH_OFF)),
+        ),
+        # Stage 2: toilet flush -> clean (sequência)
+        run_sequence(
+            service.send_msg(Message(toi_id, MessageType.FLUSH)),
+            service.send_msg(Message(toi_id, MessageType.CLEAN)),
+        ),
+    )
```
Good: Sleep program correctly encodes the SmartToilet required sequence (FLUSH -> CLEAN) using an inner `run_sequence`, while turning off Hue and Speaker in parallel in a preceding stage. This preserves the toilet dependency while allowing other safe parallelism.

mateacademy-ai-mentor commented on app/iot/service.py:
```diff
@@ -1,49 +1,122 @@
+        if device is not None:
+            await device.disconnect()
 
     def get_device(self, device_id: str) -> Device:
         return self.devices[device_id]
```
Minor note: `get_device` returns `self.devices[device_id]` and will raise `KeyError` if the id is absent. That's acceptable if callers expect that behavior, but for consistency you might either document it or use `.get` + explicit error to provide a clearer message.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -1,44 +1,16 @@
-from iot.devices import HueLightDevice, SmartSpeakerDevice, SmartToiletDevice
-from iot.message import Message, MessageType
-from iot.service import IOTService
+from iot.service import orchestrate
```
The absolute import `from iot.service import orchestrate` will work when running the project with the repository root on PYTHONPATH or when executed as a module (for example `python -m app.iot.main`). If you expect people to run this file directly from the app/iot folder, the import may fail. Consider either:
- Running as a module (`python -m app.iot.main`), or
- Using a relative import (`from .service import orchestrate`) if appropriate for your package layout.
This will avoid import errors in different run contexts.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -1,44 +1,16 @@
-    service.run_program(wake_up_program)
-    service.run_program(sleep_program)
+def _run() -> None:
+    asyncio.run(orchestrate())
```
Calling `asyncio.run(orchestrate())` is correct for starting your top-level coroutine in a normal script. Note: `asyncio.run` raises RuntimeError if called from an already-running event loop (e.g., in Jupyter notebooks or some async test harnesses). If you expect to run in such environments, consider adding an environment-aware fallback (detect running loop and use `await` or `loop.run_until_complete` accordingly) or document that this script must be run as a standalone process.

mateacademy-ai-mentor commented on app/main.py:
```diff
@@ -1,44 +1,16 @@
 if __name__ == "__main__":
     start = time.perf_counter()
-    main()
+    _run()
```
Good: the main guard calls `_run()` and the script prints elapsed time so you can attach the required screenshot. As a small improvement, consider wrapping `_run()` (or the contents of `_run`) in try/except so exceptions are clearly logged and you still get consistent elapsed-time output in case of failures. Also remember to attach the console screenshot to the PR as required by the assignment.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Educational approval: Your async refactor meets the core requirements and runs correctly; the notes below are small polish items to help you level up.

What you did well:
- Clear async device implementations using `asyncio.sleep` and consistent interfaces in `app/iot/devices.py`.
- Correct concurrency primitives: `run_sequence` preserves ordering; `run_parallel` uses `asyncio.gather` effectively; device registration/unregistration are concurrent.
- Wake-up and sleep programs are well-staged, maximizing parallelism while enforcing dependencies (Speaker on before `PLAY_SONG`; Toilet `FLUSH` -> `CLEAN`).

Improvements to consider (non-blocking):
- app/iot/service.py line where service send is defined: rename `send_msg` to `send_message` (to match the Device protocol) or add a brief comment/wrapper for API consistency. This reduces cognitive overhead for readers.
- app/iot/service.py in `get_device`: currently `self.devices[device_id]` raises `KeyError` when missing. Either document this or switch to `.get(...)` and raise a clearer `ValueError` (e.g., include the unknown id and available ids) to align with `send_msg` behavior.
- app/iot/service.py in `run_sequence`/`run_parallel`: consider renaming `*functions` to `*awaitables` or add a short docstring to clarify expected inputs.
- app/main.py import: if running as a module isn’t guaranteed, prefer a relative import (`from .service import orchestrate`) or document running with `python -m app.iot.main` to avoid import issues. Also note `asyncio.run` limitations in already-running event loops.

Overall, strong work on correctness and structure; these tweaks will improve maintainability and developer experience. Keep it up!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
